keysmash.c

```diff
@@ -1,17 +1,25 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <assert.h>
 
-char* keysmash(int length) {
-    char keys[] = { 'a', 's', 'd', 'f', 'j', 'k', 'l', 'a', 's', 'd', 'f', 'j', 'k', 'l', ';' };
-    length -= 2;
-    char* smash = (char*)malloc(sizeof(char)*1000);
+char* keysmash(unsigned length) {
+    char keys[16] = { 'a', 's', 'd', 'f', 'g', 'j', 'k', 'l', 'a', 's', 'd', 'f', 'j', 'k', 'l', ';' };
+    // Special case for this, otherwise we're writing out of bounds!
+    if (length == 1) {
+        return "a";
+    }
+    char* smash = (char*)malloc(sizeof(char)*(length+1));
     smash[0] = 'a';
     smash[1] = 's';
-    smash[2] = '\0';
-    while (length > 0) {
-        length -= 1;
-        strncat(smash, &keys[rand() % 15], 1);
+    smash[length] = '\0';
+    smash += 2;
+    length -= 2;
+    /* 0 is probably fewer cycles to check against */
+    while (length != 0) {
+        // random() is a 32 bit int. we have 16 possibilities:
+        smash[--length] = keys[random() & 15];
     }
+    smash -= 2;
```

Comment:
```suggestion

    for (int i = 2; i < length; i++) {
        // random() is a 32 bit int. we have 16 possibilities:
        smash[i] = keys[random() & 15];
    }

    smash[length] = '\0';
```

I feel like the `for` loop is a little cleaner. Plus, it's shorter.

Reply:
I could be wrong, but I believe comparisons against zero tend to be a little faster. It might just be better to do what you're saying though (but in reverse, so it can be checked `== 0`)

Reply:
This is also going to be cpu and compiler dependent. In some systems, a comparison to 0 is a one cycle op, in others, it might be multiple. You'd again need to compare profiles in the runtime container. This is premature optimisation.

Reply:
Really, in the vast majority of cases, you should optimise for readability and trust the compiler to help smooth the gaps in the asm